crates/router/src/db/merchant_account.rs

```diff
@@ -75,12 +76,29 @@ impl MerchantAccountInterface for Store {
         merchant_key_store: &domain::MerchantKeyStore,
     ) -> CustomResult<domain::MerchantAccount, errors::StorageError> {
         let conn = connection::pg_connection_write(self).await?;
-        merchant_account
+
+        let key_store = merchant_key_store
+            .clone()
             .construct_new()
             .await
-            .change_context(errors::StorageError::EncryptionError)?
-            .insert(&conn)
+            .change_context(errors::StorageError::EncryptionError)?;
+
+        let account = merchant_account
+            .construct_new()
             .await
+            .change_context(errors::StorageError::EncryptionError)?;
+
+        let final_account = conn
+            .transaction_async(|e| async move {
+                let _ = key_store.insert(&e).await;
+                account
+                    .insert(&e)
+                    .await
+                    .map_err(|_| errors::StorageError::DatabaseConnectionError)
```

Comment:
As I mentioned below if we have `From<DieselError>` for `DatabaseError` we can directly return `DatabaseError` from here and we shouldn't be ignoring errors.